--- a/drivers/aic8800/aic8800_fdrv/rwnx_main.c
+++ b/drivers/aic8800/aic8800_fdrv/rwnx_main.c
@@ -550,6 +550,7 @@
 char aic_fw_path[200];
 
 void rwnx_skb_align_8bytes(struct sk_buff *skb){
+    printk(KERN_DEBUG "aic8800: %s entered\n", __func__);
 #ifdef CONFIG_ALIGN_8BYTES
 	int align __maybe_unused;
 	u8 *data;
@@ -577,6 +578,7 @@
 void rwnx_free_cmd_array(void);
 
 void rwnx_data_dump(char* tag, void* data, unsigned long len){
+    printk(KERN_DEBUG "aic8800: %s entered\n", __func__);
 	unsigned long i = 0;
 	char* data_ = (char* )data;
 
@@ -617,6 +619,7 @@
 #define QOS_MAC_HDR_LEN 26
 
 void rwnx_frame_parser(char* tag, char* data, unsigned long len){
+    printk(KERN_DEBUG "aic8800: %s entered\n", __func__);
 	char print_data[100];
 	int print_index = 0;
 
@@ -761,6 +764,7 @@
  *********************************************************************/
 struct rwnx_sta *rwnx_get_sta(struct rwnx_hw *rwnx_hw, const u8 *mac_addr)
 {
+    printk(KERN_DEBUG "aic8800: %s entered\n", __func__);
     int i;
 
     for (i = 0; i < NX_REMOTE_STA_MAX; i++) {
@@ -774,6 +778,7 @@
 
 void rwnx_enable_wapi(struct rwnx_hw *rwnx_hw)
 {
+    printk(KERN_DEBUG "aic8800: %s entered\n", __func__);
     //cipher_suites[rwnx_hw->wiphy->n_cipher_suites] = WLAN_CIPHER_SUITE_SMS4;
     rwnx_hw->wiphy->n_cipher_suites ++;
     rwnx_hw->wiphy->flags |= WIPHY_FLAG_CONTROL_PORT_PROTOCOL;
@@ -781,6 +786,7 @@
 
 void rwnx_enable_mfp(struct rwnx_hw *rwnx_hw)
 {
+    printk(KERN_DEBUG "aic8800: %s entered\n", __func__);
     cipher_suites[rwnx_hw->wiphy->n_cipher_suites] = WLAN_CIPHER_SUITE_AES_CMAC;
     rwnx_hw->wiphy->n_cipher_suites ++;
 }
@@ -790,6 +796,7 @@
 extern u8_l vendor_extension_len;
 
 void rwnx_insert_vendor_extension_in_bcn(struct rwnx_bcn *bcn){
+    printk(KERN_DEBUG "aic8800: %s entered\n", __func__);
 	u8_l temp_ie[256];
 	u8_l vendor_extension_subelement[3] = {0x00,0x37,0x2A};
 	u8_l vendor_extension_id[2] = {0x10,0x49};
@@ -851,6 +858,7 @@
 
 u8 *rwnx_build_bcn(struct rwnx_bcn *bcn, struct cfg80211_beacon_data *new)
 {
+    printk(KERN_DEBUG "aic8800: %s entered\n", __func__);
     u8 *buf, *pos;
 
     if (new->head) {
@@ -920,6 +928,7 @@
 
 static void rwnx_del_bcn(struct rwnx_bcn *bcn)
 {
+    printk(KERN_DEBUG "aic8800: %s entered\n", __func__);
     if (bcn->head) {
         kfree(bcn->head);
         bcn->head = NULL;
@@ -948,6 +957,7 @@
 void rwnx_chanctx_link(struct rwnx_vif *vif, u8 ch_idx,
                        struct cfg80211_chan_def *chandef)
 {
+    printk(KERN_DEBUG "aic8800: %s entered\n", __func__);
     struct rwnx_chanctx *ctxt;
 
     if (ch_idx >= NX_CHAN_CTXT_CNT) {
@@ -975,6 +985,7 @@
  */
 void rwnx_chanctx_unlink(struct rwnx_vif *vif)
 {
+    printk(KERN_DEBUG "aic8800: %s entered\n", __func__);
     struct rwnx_chanctx *ctxt;
 
     if (vif->ch_index == RWNX_CH_NOT_SET)
@@ -1005,6 +1016,7 @@
 
 int rwnx_chanctx_valid(struct rwnx_hw *rwnx_hw, u8 ch_idx)
 {
+    printk(KERN_DEBUG "aic8800: %s entered\n", __func__);
     if (ch_idx >= NX_CHAN_CTXT_CNT ||
         rwnx_hw->chanctx_table[ch_idx].chan_def.chan == NULL) {
         return 0;
@@ -1015,6 +1027,7 @@
 
 static void rwnx_del_csa(struct rwnx_vif *vif)
 {
+    printk(KERN_DEBUG "aic8800: %s entered\n", __func__);
     //struct rwnx_hw *rwnx_hw = vif->rwnx_hw;
     struct rwnx_csa *csa = vif->ap.csa;
 
@@ -1122,6 +1135,7 @@
  */
 void rwnx_external_auth_enable(struct rwnx_vif *vif)
 {
+    printk(KERN_DEBUG "aic8800: %s entered\n", __func__);
     vif->sta.external_auth = true;
     rwnx_txq_unk_vif_init(vif);
     rwnx_txq_start(rwnx_txq_vif_get(vif, NX_UNK_TXQ_TYPE), 0);
@@ -1134,6 +1148,7 @@
  */
 void rwnx_external_auth_disable(struct rwnx_vif *vif)
 {
+    printk(KERN_DEBUG "aic8800: %s entered\n", __func__);
     if (!vif->sta.external_auth)
         return;
 
@@ -1156,6 +1171,7 @@
  */
 void rwnx_update_mesh_power_mode(struct rwnx_vif *vif)
 {
+    printk(KERN_DEBUG "aic8800: %s entered\n", __func__);
     enum nl80211_mesh_power_mode mesh_pm;
     struct rwnx_sta *sta;
     struct mesh_config mesh_conf;
@@ -1313,6 +1329,7 @@
  */
 static int rwnx_open(struct net_device *dev)
 {
+    printk(KERN_DEBUG "aic8800: %s entered\n", __func__);
     struct rwnx_vif *rwnx_vif = netdev_priv(dev);
     struct rwnx_hw *rwnx_hw = rwnx_vif->rwnx_hw;
     struct mm_add_if_cfm add_if_cfm;
@@ -1330,6 +1347,7 @@
 		AICWFDBG(LOGDEBUG, "%s waiting for rwnx_close \r\n", __func__);
 		waiting_counter--;
 		if(waiting_counter == 0){
+    printk(KERN_DEBUG "aic8800: %s entered\n", __func__);
 			AICWFDBG(LOGERROR, "%s error waiting for close time out \r\n", __func__);
 			break;
 		}
@@ -1684,6 +1702,7 @@
 static int rwnx_do_ioctl(struct net_device *net, struct ifreq *req, int cmd)
 #endif
 {
+    printk(KERN_DEBUG "aic8800: %s entered\n", __func__);
     int ret = 0;
     ///TODO: add ioctl command handler later
     switch(cmd)
@@ -1748,6 +1767,7 @@
  */
 static int rwnx_set_mac_address(struct net_device *dev, void *addr)
 {
+    printk(KERN_DEBUG "aic8800: %s entered\n", __func__);
     struct sockaddr *sa = addr;
     int ret;
     struct rwnx_vif *rwnx_vif = netdev_priv(dev);
@@ -1797,6 +1817,7 @@
 
 static void rwnx_netdev_setup(struct net_device *dev)
 {
+    printk(KERN_DEBUG "aic8800: %s entered\n", __func__);
     ether_setup(dev);
     dev->priv_flags &= ~IFF_TX_SKB_SHARING;
     dev->netdev_ops = &rwnx_netdev_ops;
@@ -1835,6 +1856,7 @@
                                                enum nl80211_iftype type,
                                                struct vif_params *params)
 {
+    printk(KERN_DEBUG "aic8800: %s entered\n", __func__);
     struct net_device *ndev;
     struct rwnx_vif *vif;
     int min_idx, max_idx;
@@ -2029,6 +2051,7 @@
 void aicwf_p2p_alive_timeout(struct timer_list *t)
 #endif
 {
+    printk(KERN_DEBUG "aic8800: %s entered\n", __func__);
     struct rwnx_hw *rwnx_hw;
     struct rwnx_vif *rwnx_vif;
     struct rwnx_vif *rwnx_vif1, *tmp;
@@ -2254,6 +2277,7 @@
                                                enum nl80211_iftype type,
                                                struct vif_params *params)
 {
+    printk(KERN_DEBUG "aic8800: %s entered\n", __func__);
     struct wireless_dev *wdev = NULL;
     struct rwnx_vif *vif;
     int min_idx, max_idx;
@@ -2341,6 +2365,7 @@
                                           struct rwnx_vif *rwnx_vif, u8 *addr,
                                           __le16 fc, bool ap)
 {
+    printk(KERN_DEBUG "aic8800: %s entered\n", __func__);
     if (ap) {
         /* only deauth, disassoc and action are bufferable MMPDUs */
         bool bufferable = ieee80211_is_deauth(fc) ||
@@ -2398,6 +2423,7 @@
                                                     enum nl80211_iftype type,
                                                     struct vif_params *params)
 {
+    printk(KERN_DEBUG "aic8800: %s entered\n", __func__);
     struct rwnx_hw *rwnx_hw = wiphy_priv(wiphy);
     struct wireless_dev *wdev;
 #if (LINUX_VERSION_CODE < KERNEL_VERSION(4, 1, 0))
@@ -2423,6 +2449,7 @@
  */
 static int rwnx_cfg80211_del_iface(struct wiphy *wiphy, struct wireless_dev *wdev)
 {
+    printk(KERN_DEBUG "aic8800: %s entered\n", __func__);
 #if (LINUX_VERSION_CODE >= KERNEL_VERSION(3, 6, 0))
     struct net_device *dev = wdev->netdev;
     struct rwnx_hw *rwnx_hw = wiphy_priv(wiphy);
@@ -2474,7 +2501,7 @@
     rwnx_vif->ndev = NULL;
 
     /* Clear the priv in adapter */
-    dev->ieee80211_ptr = NULL;
+    //dev->ieee80211_ptr = NULL;
 
     return 0;
 }
@@ -2620,6 +2647,7 @@
 
 static int rwnx_cfgp2p_start_p2p_device(struct wiphy *wiphy, struct wireless_dev *wdev)
 {
+    printk(KERN_DEBUG "aic8800: %s entered\n", __func__);
     int ret = 0;
 
     //do nothing
@@ -2630,6 +2658,7 @@
 
 static void rwnx_cfgp2p_stop_p2p_device(struct wiphy *wiphy, struct wireless_dev *wdev)
 {
+    printk(KERN_DEBUG "aic8800: %s entered\n", __func__);
 #if 0
     int ret = 0;
     struct bcm_cfg80211 *cfg = wiphy_priv(wiphy);
@@ -2706,6 +2735,7 @@
 }
 
 int rwnx_send_check_p2p(struct cfg80211_scan_request *param){
+    printk(KERN_DEBUG "aic8800: %s entered\n", __func__);
 	int index = (u8)min_t(int, SCAN_SSID_MAX, param->n_ssids);
 	int i = 0;
 
@@ -2731,6 +2761,7 @@
 	#endif
                               struct cfg80211_scan_request *request)
 {
+    printk(KERN_DEBUG "aic8800: %s entered\n", __func__);
     struct rwnx_hw *rwnx_hw = wiphy_priv(wiphy);
 #if LINUX_VERSION_CODE >= KERNEL_VERSION(3, 6, 0)
     struct rwnx_vif *rwnx_vif = container_of(request->wdev, struct rwnx_vif,
@@ -2795,6 +2826,7 @@
                                  u8 key_index, bool pairwise, const u8 *mac_addr,
                                  struct key_params *params)
 {
+    printk(KERN_DEBUG "aic8800: %s entered\n", __func__);
     struct rwnx_hw *rwnx_hw = wiphy_priv(wiphy);
     struct rwnx_vif *vif = netdev_priv(netdev);
     int i, error = 0;
@@ -2893,6 +2925,7 @@
                                  void *cookie,
                                  void (*callback)(void *cookie, struct key_params*))
 {
+    printk(KERN_DEBUG "aic8800: %s entered\n", __func__);
     RWNX_DBG(RWNX_FN_ENTRY_STR);
 
     return -1;
@@ -2910,6 +2943,7 @@
 
                                  u8 key_index, bool pairwise, const u8 *mac_addr)
 {
+    printk(KERN_DEBUG "aic8800: %s entered\n", __func__);
     struct rwnx_hw *rwnx_hw = wiphy_priv(wiphy);
     struct rwnx_vif *vif = netdev_priv(netdev);
     int error;
@@ -2950,6 +2984,7 @@
 #endif
                                          u8 key_index, bool unicast, bool multicast)
 {
+    printk(KERN_DEBUG "aic8800: %s entered\n", __func__);
     RWNX_DBG(RWNX_FN_ENTRY_STR);
 
     return 0;
@@ -2965,6 +3000,7 @@
 #endif
                                               u8 key_index)
 {
+    printk(KERN_DEBUG "aic8800: %s entered\n", __func__);
     return 0;
 }
 
@@ -2979,6 +3015,7 @@
 static int rwnx_cfg80211_connect(struct wiphy *wiphy, struct net_device *dev,
                                  struct cfg80211_connect_params *sme)
 {
+    printk(KERN_DEBUG "aic8800: %s entered\n", __func__);
     struct rwnx_hw *rwnx_hw = wiphy_priv(wiphy);
     struct rwnx_vif *rwnx_vif = netdev_priv(dev);
     struct sm_connect_cfm sm_connect_cfm;
@@ -3104,6 +3141,7 @@
 static int rwnx_cfg80211_disconnect(struct wiphy *wiphy, struct net_device *dev,
                                     u16 reason_code)
 {
+    printk(KERN_DEBUG "aic8800: %s entered\n", __func__);
     struct rwnx_hw *rwnx_hw = wiphy_priv(wiphy);
     struct rwnx_vif *rwnx_vif = netdev_priv(dev);
 
@@ -3158,6 +3196,7 @@
                         )
 #endif
 {
+    printk(KERN_DEBUG "aic8800: %s entered\n", __func__);
 
 	struct rwnx_hw *rwnx_hw = wiphy_priv(wiphy);
 	//struct rwnx_vif *rwnx_vif = netdev_priv(dev);
@@ -3177,6 +3216,7 @@
                              struct cfg80211_sched_scan_request *request)
 
 {
+    printk(KERN_DEBUG "aic8800: %s entered\n", __func__);
     struct rwnx_hw *rwnx_hw = wiphy_priv(wiphy);
 	struct rwnx_vif *rwnx_vif = netdev_priv(dev);
     struct cfg80211_scan_request *scan_request = NULL;
@@ -3259,6 +3299,7 @@
 static int rwnx_cfg80211_external_auth(struct wiphy *wiphy, struct net_device *dev,
                                        struct cfg80211_external_auth_params *params)
 {
+    printk(KERN_DEBUG "aic8800: %s entered\n", __func__);
     struct rwnx_hw *rwnx_hw = wiphy_priv(wiphy);
     struct rwnx_vif *rwnx_vif = netdev_priv(dev);
 
@@ -3284,6 +3325,7 @@
 #endif
 	struct station_parameters *params)
 {
+    printk(KERN_DEBUG "aic8800: %s entered\n", __func__);
     struct rwnx_hw *rwnx_hw = wiphy_priv(wiphy);
     struct rwnx_vif *rwnx_vif = netdev_priv(dev);
     struct me_sta_add_cfm me_sta_add_cfm;
@@ -3437,6 +3479,7 @@
 #endif
 )
 {
+    printk(KERN_DEBUG "aic8800: %s entered\n", __func__);
     struct rwnx_hw *rwnx_hw = wiphy_priv(wiphy);
     struct rwnx_vif *rwnx_vif = netdev_priv(dev);
     struct rwnx_sta *cur, *tmp;
@@ -3563,6 +3606,7 @@
 
 void apm_staloss_work_process(struct work_struct *work)
 {
+    printk(KERN_DEBUG "aic8800: %s entered\n", __func__);
 	struct rwnx_hw *rwnx_hw = container_of(work, struct rwnx_hw, apmStalossWork);
 	struct rwnx_sta *cur, *tmp;
 	int error = 0;
@@ -3677,6 +3721,7 @@
 
 void apm_probe_sta_work_process(struct work_struct *work)
 {
+    printk(KERN_DEBUG "aic8800: %s entered\n", __func__);
        struct apm_probe_sta *probe_sta = container_of(work, struct apm_probe_sta, apmprobestaWork);
        struct rwnx_vif *rwnx_vif = container_of(probe_sta, struct rwnx_vif, sta_probe);
        bool found = false;
@@ -3725,6 +3770,7 @@
 #endif
 	struct station_parameters *params)
 {
+    printk(KERN_DEBUG "aic8800: %s entered\n", __func__);
     struct rwnx_hw *rwnx_hw = wiphy_priv(wiphy);
     struct rwnx_vif *vif = netdev_priv(dev);
     struct rwnx_sta *sta;
@@ -3884,6 +3930,7 @@
 static int rwnx_cfg80211_start_ap(struct wiphy *wiphy, struct net_device *dev,
                                   struct cfg80211_ap_settings *settings)
 {
+    printk(KERN_DEBUG "aic8800: %s entered\n", __func__);
     struct rwnx_hw *rwnx_hw = wiphy_priv(wiphy);
     struct rwnx_vif *rwnx_vif = netdev_priv(dev);
     struct apm_start_cfm apm_start_cfm;
@@ -4024,6 +4071,7 @@
                                        struct cfg80211_beacon_data *info)
 #endif
 {
+    printk(KERN_DEBUG "aic8800: %s entered\n", __func__);
     struct rwnx_hw *rwnx_hw = wiphy_priv(wiphy);
     struct rwnx_vif *vif = netdev_priv(dev);
     struct rwnx_bcn *bcn = &vif->ap.bcn;
@@ -4075,6 +4123,7 @@
 #endif
 
 {
+    printk(KERN_DEBUG "aic8800: %s entered\n", __func__);
     struct rwnx_hw *rwnx_hw = wiphy_priv(wiphy);
     struct rwnx_vif *rwnx_vif = netdev_priv(dev);
     struct rwnx_sta *sta;
@@ -4139,6 +4188,7 @@
 cfg80211_chandef_identical(const struct cfg80211_chan_def *chandef1,
 			   const struct cfg80211_chan_def *chandef2)
 {
+    printk(KERN_DEBUG "aic8800: %s entered\n", __func__);
 	return (chandef1->chan == chandef2->chan &&
 		chandef1->width == chandef2->width &&
 		chandef1->center_freq1 == chandef2->center_freq1 &&
@@ -4154,6 +4204,7 @@
                                              struct cfg80211_chan_def *chandef)
 #endif
 {
+    printk(KERN_DEBUG "aic8800: %s entered\n", __func__);
     struct rwnx_hw *rwnx_hw = wiphy_priv(wiphy);
     struct rwnx_vif *rwnx_vif;
     struct me_config_monitor_cfm cfm;
@@ -4229,6 +4280,7 @@
 int rwnx_cfg80211_probe_client(struct wiphy *wiphy, struct net_device *dev,
             const u8 *peer, u64 *cookie)
 {
+    printk(KERN_DEBUG "aic8800: %s entered\n", __func__);
     //struct rwnx_hw *rwnx_hw = wiphy_priv(wiphy);
     struct rwnx_vif *vif = netdev_priv(dev);
     struct rwnx_sta *sta = NULL;
@@ -4271,6 +4323,7 @@
 #endif
                    u16 frame_type, bool reg)
 {
+    printk(KERN_DEBUG "aic8800: %s entered\n", __func__);
 }
 
 /**
@@ -4285,6 +4338,7 @@
 static int rwnx_cfg80211_set_wiphy_params(struct wiphy *wiphy, u32 changed)
 #endif
 {
+    printk(KERN_DEBUG "aic8800: %s entered\n", __func__);
     return 0;
 }
 
@@ -4305,6 +4359,7 @@
 #endif
                                       enum nl80211_tx_power_setting type, int mbm)
 {
+    printk(KERN_DEBUG "aic8800: %s entered\n", __func__);
     #if LINUX_VERSION_CODE < KERNEL_VERSION(3, 8, 0)
     struct wireless_dev *wdev = NULL;
     #endif
@@ -4343,6 +4398,7 @@
                                         struct net_device *dev,
                                         bool enabled, int timeout)
 {
+    printk(KERN_DEBUG "aic8800: %s entered\n", __func__);
 #if 0
 
     struct rwnx_hw *rwnx_hw = wiphy_priv(wiphy);
@@ -4373,6 +4429,7 @@
 static int rwnx_cfg80211_set_txq_params(struct wiphy *wiphy, struct net_device *dev,
                                         struct ieee80211_txq_params *params)
 {
+    printk(KERN_DEBUG "aic8800: %s entered\n", __func__);
     struct rwnx_hw *rwnx_hw = wiphy_priv(wiphy);
     struct rwnx_vif *rwnx_vif = netdev_priv(dev);
     u8 hw_queue, aifs, cwmin, cwmax;
@@ -4422,6 +4479,7 @@
                             #endif
                                 unsigned int duration, u64 *cookie, bool mgmt_roc_flag)
 {
+    printk(KERN_DEBUG "aic8800: %s entered\n", __func__);
     struct rwnx_hw *rwnx_hw = wiphy_priv(wiphy);
 #if LINUX_VERSION_CODE >= KERNEL_VERSION(3, 6, 0)
     struct rwnx_vif *rwnx_vif = container_of(wdev, struct rwnx_vif, wdev);
@@ -4561,6 +4619,7 @@
                             #endif
                                 unsigned int duration, u64 *cookie)
 {
+    printk(KERN_DEBUG "aic8800: %s entered\n", __func__);
 	return rwnx_cfg80211_remain_on_channel_(wiphy,
                             #if LINUX_VERSION_CODE >= KERNEL_VERSION(3, 6, 0)
                                 wdev,
@@ -4587,6 +4646,7 @@
                                             #endif
                                                   u64 cookie)
 {
+    printk(KERN_DEBUG "aic8800: %s entered\n", __func__);
     struct rwnx_hw *rwnx_hw = wiphy_priv(wiphy);
 #ifdef CREATE_TRACE_POINTS
 #if LINUX_VERSION_CODE < KERNEL_VERSION(3, 6, 0)
@@ -4619,6 +4679,7 @@
 static int rwnx_cfg80211_dump_survey(struct wiphy *wiphy, struct net_device *netdev,
                                      int idx, struct survey_info *info)
 {
+    printk(KERN_DEBUG "aic8800: %s entered\n", __func__);
     struct rwnx_hw *rwnx_hw = wiphy_priv(wiphy);
     struct ieee80211_supported_band *sband;
     struct rwnx_survey_info *rwnx_survey;
@@ -4690,6 +4751,7 @@
 #endif
                                                     struct cfg80211_chan_def *chandef)
 {
+    printk(KERN_DEBUG "aic8800: %s entered\n", __func__);
     struct rwnx_hw *rwnx_hw = wiphy_priv(wiphy);
     struct rwnx_vif *rwnx_vif = container_of(wdev, struct rwnx_vif, wdev);
     struct rwnx_chanctx *ctxt;
@@ -4766,6 +4828,7 @@
                                  u64 *cookie)
 #endif /* LINUX_VERSION_CODE >= KERNEL_VERSION(3, 14, 0) */
 {
+    printk(KERN_DEBUG "aic8800: %s entered\n", __func__);
     struct rwnx_hw *rwnx_hw = wiphy_priv(wiphy);
 #if (LINUX_VERSION_CODE < KERNEL_VERSION(3, 6, 0))
     struct rwnx_vif *rwnx_vif = netdev_priv(dev);
@@ -4892,7 +4955,7 @@
 /**
  * @start_radar_detection: Start radar detection in the driver.
  */
-static
+static __maybe_unused
 int rwnx_cfg80211_start_radar_detection(struct wiphy *wiphy,
                                         struct net_device *dev,
                                         struct cfg80211_chan_def *chandef
@@ -4940,6 +5003,7 @@
                             struct net_device *dev,
                             struct cfg80211_update_ft_ies_params *ftie)
 {
+    printk(KERN_DEBUG "aic8800: %s entered\n", __func__);
     return 0;
 }
 
@@ -4951,6 +5015,7 @@
                                   struct net_device *dev,
                                   int32_t rssi_thold, uint32_t rssi_hyst)
 {
+    printk(KERN_DEBUG "aic8800: %s entered\n", __func__);
     struct rwnx_hw *rwnx_hw = wiphy_priv(wiphy);
     struct rwnx_vif *rwnx_vif = netdev_priv(dev);
 
@@ -5312,6 +5377,7 @@
 int rwnx_cfg80211_change_bss(struct wiphy *wiphy, struct net_device *dev,
                              struct bss_parameters *params)
 {
+    printk(KERN_DEBUG "aic8800: %s entered\n", __func__);
     struct rwnx_vif *rwnx_vif = netdev_priv(dev);
     int res =  -EOPNOTSUPP;
 
@@ -5440,6 +5506,7 @@
 int rwnx_fill_station_info(struct rwnx_sta *sta, struct rwnx_vif *vif,
 								  struct station_info *sinfo, u8 *phymode, u32 *tx_phyrate, u32 *rx_phyrate)
 {
+    printk(KERN_DEBUG "aic8800: %s entered\n", __func__);
 	struct rwnx_sta_stats *stats = &sta->stats;
 	struct rx_vector_1 *rx_vect1 = &stats->last_rx.rx_vect1;
 	union rwnx_rate_ctrl_info *rate_info;
@@ -5690,6 +5757,7 @@
 #endif
 	struct station_info *sinfo)
 {
+    printk(KERN_DEBUG "aic8800: %s entered\n", __func__);
     struct rwnx_vif *vif = netdev_priv(dev);
     struct rwnx_sta *sta = NULL;
 
@@ -5726,6 +5794,7 @@
 static int rwnx_cfg80211_dump_station(struct wiphy *wiphy, struct net_device *dev,
                                       int idx, u8 *mac, struct station_info *sinfo)
 {
+    printk(KERN_DEBUG "aic8800: %s entered\n", __func__);
     struct rwnx_vif *rwnx_vif = netdev_priv(dev);
     //struct rwnx_hw *rwnx_hw = wiphy_priv(wiphy);
     struct rwnx_sta *sta_iter, *sta = NULL;
@@ -5811,6 +5880,7 @@
 #endif
 )
 {
+    printk(KERN_DEBUG "aic8800: %s entered\n", __func__);
     struct rwnx_vif *rwnx_vif = netdev_priv(dev);
     struct rwnx_hw *rwnx_hw = wiphy_priv(wiphy);
     struct mesh_path_update_cfm cfm;
@@ -5832,6 +5902,7 @@
 #endif
 )
 {
+    printk(KERN_DEBUG "aic8800: %s entered\n", __func__);
     struct rwnx_vif *rwnx_vif = netdev_priv(dev);
     struct rwnx_hw *rwnx_hw = wiphy_priv(wiphy);
     struct mesh_path_update_cfm cfm;
@@ -5855,6 +5926,7 @@
 #endif
 )
 {
+    printk(KERN_DEBUG "aic8800: %s entered\n", __func__);
     struct rwnx_vif *rwnx_vif = netdev_priv(dev);
     struct rwnx_hw *rwnx_hw = wiphy_priv(wiphy);
     struct mesh_path_update_cfm cfm;
@@ -5909,6 +5981,7 @@
                                     int idx, u8 *dst, u8 *next_hop,
                                     struct mpath_info *pinfo)
 {
+    printk(KERN_DEBUG "aic8800: %s entered\n", __func__);
     struct rwnx_vif *rwnx_vif = netdev_priv(dev);
     struct rwnx_mesh_path *mesh_path = NULL;
     struct rwnx_mesh_path *cur;
@@ -5949,6 +6022,7 @@
 static int rwnx_cfg80211_get_mpp(struct wiphy *wiphy, struct net_device *dev,
                                  u8 *dst, u8 *mpp, struct mpath_info *pinfo)
 {
+    printk(KERN_DEBUG "aic8800: %s entered\n", __func__);
     struct rwnx_vif *rwnx_vif = netdev_priv(dev);
     struct rwnx_mesh_proxy *mesh_proxy = NULL;
     struct rwnx_mesh_proxy *cur;
@@ -5988,6 +6062,7 @@
 static int rwnx_cfg80211_dump_mpp(struct wiphy *wiphy, struct net_device *dev,
                                   int idx, u8 *dst, u8 *mpp, struct mpath_info *pinfo)
 {
+    printk(KERN_DEBUG "aic8800: %s entered\n", __func__);
     struct rwnx_vif *rwnx_vif = netdev_priv(dev);
     struct rwnx_mesh_proxy *mesh_proxy = NULL;
     struct rwnx_mesh_proxy *cur;
@@ -6031,6 +6106,7 @@
 static int rwnx_cfg80211_get_mesh_config(struct wiphy *wiphy, struct net_device *dev,
                                          struct mesh_config *conf)
 {
+    printk(KERN_DEBUG "aic8800: %s entered\n", __func__);
     struct rwnx_vif *rwnx_vif = netdev_priv(dev);
 
     if (RWNX_VIF_TYPE(rwnx_vif) != NL80211_IFTYPE_MESH_POINT)
@@ -6045,6 +6121,7 @@
 static int rwnx_cfg80211_update_mesh_config(struct wiphy *wiphy, struct net_device *dev,
                                             u32 mask, const struct mesh_config *nconf)
 {
+    printk(KERN_DEBUG "aic8800: %s entered\n", __func__);
     struct rwnx_vif *rwnx_vif = netdev_priv(dev);
     struct rwnx_hw *rwnx_hw = wiphy_priv(wiphy);
     struct mesh_update_cfm cfm;
@@ -6082,6 +6159,7 @@
 static int rwnx_cfg80211_join_mesh(struct wiphy *wiphy, struct net_device *dev,
                                    const struct mesh_config *conf, const struct mesh_setup *setup)
 {
+    printk(KERN_DEBUG "aic8800: %s entered\n", __func__);
     struct rwnx_hw *rwnx_hw = wiphy_priv(wiphy);
     struct rwnx_vif *rwnx_vif = netdev_priv(dev);
     struct mesh_start_cfm mesh_start_cfm;
@@ -6165,6 +6243,7 @@
  */
 static int rwnx_cfg80211_leave_mesh(struct wiphy *wiphy, struct net_device *dev)
 {
+    printk(KERN_DEBUG "aic8800: %s entered\n", __func__);
     struct rwnx_hw *rwnx_hw = wiphy_priv(wiphy);
     struct rwnx_vif *rwnx_vif = netdev_priv(dev);
     struct mesh_stop_cfm mesh_stop_cfm;
@@ -6267,6 +6346,7 @@
  *********************************************************************/
 static void rwnx_wdev_unregister(struct rwnx_hw *rwnx_hw)
 {
+    printk(KERN_DEBUG "aic8800: %s entered\n", __func__);
     struct rwnx_vif *rwnx_vif, *tmp;
 
     rtnl_lock();
@@ -6282,6 +6362,7 @@
 
 static void rwnx_set_vers(struct rwnx_hw *rwnx_hw)
 {
+    printk(KERN_DEBUG "aic8800: %s entered\n", __func__);
     u32 vers = rwnx_hw->version_cfm.version_lmac;
 
     RWNX_DBG(RWNX_FN_ENTRY_STR);
@@ -6295,6 +6376,7 @@
 static void rwnx_reg_notifier(struct wiphy *wiphy,
                               struct regulatory_request *request)
 {
+    printk(KERN_DEBUG "aic8800: %s entered\n", __func__);
     struct rwnx_hw *rwnx_hw = wiphy_priv(wiphy);
 
     printk("%s Enter\r\n", __func__);
@@ -6319,6 +6401,7 @@
 
 static void rwnx_enable_mesh(struct rwnx_hw *rwnx_hw)
 {
+    printk(KERN_DEBUG "aic8800: %s entered\n", __func__);
     struct wiphy *wiphy = rwnx_hw->wiphy;
 
     if (!rwnx_mod_params.mesh)
@@ -8221,6 +8304,7 @@
 };
 
 static void system_config_8800(struct rwnx_hw *rwnx_hw){
+    printk(KERN_DEBUG "aic8800: %s entered\n", __func__);
     int syscfg_num;
     int ret, cnt;
     const u32 mem_addr = 0x40500000;
@@ -8299,6 +8383,7 @@
 
 static void system_config(struct rwnx_hw *rwnx_hw)
 {
+    printk(KERN_DEBUG "aic8800: %s entered\n", __func__);
 	if(rwnx_hw->usbdev->chipid == PRODUCT_ID_AIC8801){
 		system_config_8800(rwnx_hw);
 	}else if(rwnx_hw->usbdev->chipid == PRODUCT_ID_AIC8800DC ||
@@ -8309,6 +8394,7 @@
 
 static int wdt_config(struct rwnx_hw *rwnx_hw)
 {
+    printk(KERN_DEBUG "aic8800: %s entered\n", __func__);
     int ret = 0;
     ret = rwnx_send_dbg_mem_mask_write_req(rwnx_hw,
                 wdt_tbl_masked[0][0], wdt_tbl_masked[0][1], wdt_tbl_masked[0][2]);
@@ -8331,6 +8417,7 @@
 #if 0
 static int start_from_bootrom(struct rwnx_hw *rwnx_hw)
 {
+    printk(KERN_DEBUG "aic8800: %s entered\n", __func__);
     int ret = 0;
 
 	/* memory access */
@@ -8377,6 +8464,7 @@
 
 static void aicwf_usb_config(struct rwnx_hw *rwnx_hw)
 {
+    printk(KERN_DEBUG "aic8800: %s entered\n", __func__);
     int usbcfg_num = 0;
     int ret = 0, cnt = 0;
     struct dbg_mem_read_cfm rd_mem_addr_cfm;
@@ -8437,6 +8525,7 @@
 
 
 int rwnx_ic_system_init(struct rwnx_hw *rwnx_hw){
+    printk(KERN_DEBUG "aic8800: %s entered\n", __func__);
 
 	if(rwnx_hw->usbdev->chipid == PRODUCT_ID_AIC8801){
 		system_config(rwnx_hw);
@@ -8476,6 +8565,7 @@
 
 
 int rwnx_ic_rf_init(struct rwnx_hw *rwnx_hw){
+    printk(KERN_DEBUG "aic8800: %s entered\n", __func__);
 	struct mm_set_rf_calib_cfm cfm;
 	int ret = 0;
 #ifdef CONFIG_5M10M
@@ -8534,6 +8624,7 @@
 extern char default_ccode[];
 int rwnx_cfg80211_init(struct rwnx_plat *rwnx_plat, void **platform_data)
 {
+    printk(KERN_DEBUG "aic8800: %s entered\n", __func__);
     struct rwnx_hw *rwnx_hw;
     struct rwnx_conf_file init_conf;
     int ret = 0;
@@ -9021,6 +9112,7 @@
 
 void rwnx_cfg80211_deinit(struct rwnx_hw *rwnx_hw)
 {
+    printk(KERN_DEBUG "aic8800: %s entered\n", __func__);
     struct mm_set_stack_start_cfm set_start_cfm;
     struct defrag_ctrl_info *defrag_ctrl = NULL;
 
@@ -9096,6 +9188,7 @@
 
 static void aicsmac_driver_register(void)
 {
+    printk(KERN_DEBUG "aic8800: %s entered\n", __func__);
 #ifdef AICWF_SDIO_SUPPORT
     aicwf_sdio_register();
 #endif
@@ -9115,12 +9208,14 @@
 
 void aicwf_hostif_ready(void)
 {
+    printk(KERN_DEBUG "aic8800: %s entered\n", __func__);
 	g_rwnx_plat->enabled = true;
 	complete(&hostif_register_done);
 }
 
 static int __init rwnx_mod_init(void)
 {
+    printk(KERN_DEBUG "aic8800: %s entered\n", __func__);
 
     RWNX_DBG(RWNX_FN_ENTRY_STR);
     rwnx_print_version();
@@ -9159,6 +9254,7 @@
  */
 static void __exit rwnx_mod_exit(void)
 {
+    printk(KERN_DEBUG "aic8800: %s entered\n", __func__);
     RWNX_DBG(RWNX_FN_ENTRY_STR);
 
 #ifdef AICWF_PCIE_SUPPORT
