--- a/drivers/aic8800/aic8800_fdrv/rwnx_defs.h
+++ b/drivers/aic8800/aic8800_fdrv/rwnx_defs.h
@@ -215,33 +215,33 @@
 
 #ifdef CONFIG_BAND_STEERING
 enum band_type {
-	BAND_ON_24G = 0,
-	BAND_ON_5G = 1,
-	BAND_ON_60G = 2,
-	BAND_ON_6G = 3,
-	BAND_MAX,
+        BAND_ON_24G = 0,
+        BAND_ON_5G = 1,
+        BAND_ON_60G = 2,
+        BAND_ON_6G = 3,
+        BAND_MAX,
 };
 
 enum WIFI_FRAME_TYPE {
-	WIFI_MGT_TYPE  = (0),
+        WIFI_MGT_TYPE  = (0),
 };
 
 enum WIFI_FRAME_SUBTYPE {
-	WIFI_ASSOCREQ       = (0 | WIFI_MGT_TYPE),
-	WIFI_PROBEREQ       = (BIT(6) | WIFI_MGT_TYPE),
-	WIFI_AUTH           = (BIT(7) | BIT(5) | BIT(4) | WIFI_MGT_TYPE),
+        WIFI_ASSOCREQ       = (0 | WIFI_MGT_TYPE),
+        WIFI_PROBEREQ       = (BIT(6) | WIFI_MGT_TYPE),
+        WIFI_AUTH           = (BIT(7) | BIT(5) | BIT(4) | WIFI_MGT_TYPE),
 };
 
 struct tmp_feature_sta {
-	u8_l sta_idx;
-	u8_l supported_band;
+        u8_l sta_idx;
+        u8_l supported_band;
 };
 
 #define MAX_PENDING_PROBES 3
 struct ap_probe_rsp {
-	u8_l da[6];
-	struct work_struct rsp_work;
-	bool in_use;
+        u8_l da[6];
+        struct work_struct rsp_work;
+        bool in_use;
 };
 #endif
 
@@ -252,13 +252,13 @@
  * TDLS capabililites to be enabled in the 5th byte of the
  * @WLAN_EID_EXT_CAPABILITY information element
  */
-#define WLAN_EXT_CAPA5_TDLS_ENABLED	BIT(5)
-#define WLAN_EXT_CAPA5_TDLS_PROHIBITED	BIT(6)
+#define WLAN_EXT_CAPA5_TDLS_ENABLED     BIT(5)
+#define WLAN_EXT_CAPA5_TDLS_PROHIBITED  BIT(6)
 
-#define WLAN_EXT_CAPA8_OPMODE_NOTIF	BIT(6)
+#define WLAN_EXT_CAPA8_OPMODE_NOTIF     BIT(6)
 
 /* TDLS specific payload type in the LLC/SNAP header */
-#define WLAN_TDLS_SNAP_RFTYPE	0x2
+#define WLAN_TDLS_SNAP_RFTYPE   0x2
 
 #endif
 
@@ -428,13 +428,13 @@
             bool external_auth;  /* Indicate if external authentication is in progress */
             u32 group_cipher_type;
             u32 paired_cipher_type;
-			//connected network info start
-			char ssid[33];//ssid max is 32, but this has one spare for '\0'
-			int ssid_len;
-			u8 bssid[ETH_ALEN];
-			u32 conn_owner_nlportid;
-			bool is_roam;
-			//connected network info end
+                        //connected network info start
+                        char ssid[33];//ssid max is 32, but this has one spare for '\0'
+                        int ssid_len;
+                        u8 bssid[ETH_ALEN];
+                        u32 conn_owner_nlportid;
+                        bool is_roam;
+                        //connected network info end
         } sta;
         struct
         {
@@ -443,7 +443,7 @@
             struct rwnx_bcn bcn;       /* beacon */
             u8 bcmc_index;             /* Index of the BCMC sta to use */
             #if (defined CONFIG_HE_FOR_OLD_KERNEL) || (defined CONFIG_VHT_FOR_OLD_KERNEL)
-			u8 aic_index;
+                        u8 aic_index;
             #endif
             struct rwnx_csa *csa;
 
@@ -453,10 +453,10 @@
                                             message */
             int generation;              /* Increased each time the list of Mesh Paths is updated */
 #ifdef CONFIG_BAND_STEERING
-			u8_l tmp_sta_idx;
-			enum band_type band;
-			u32_l freq;
-			bool start;
+                        u8_l tmp_sta_idx;
+                        enum band_type band;
+                        u32_l freq;
+                        bool start;
 #endif
             enum nl80211_mesh_power_mode mesh_pm; /* mesh power save mode currently set in firmware */
             enum nl80211_mesh_power_mode next_mesh_pm; /* mesh power save mode for next peer */
@@ -468,30 +468,32 @@
         } ap_vlan;
     };
 
-	u8_l key_has_add;
-	u8_l is_p2p_vif;
-	struct apm_probe_sta sta_probe;
+        u8_l key_has_add;
+        u8_l is_p2p_vif;
+        bool deleted;                   /* Flag to prevent double list_del and
+                                           use-after-free in race conditions */
+        struct apm_probe_sta sta_probe;
 
     #ifdef CONFIG_BR_SUPPORT
-	spinlock_t			    br_ext_lock;
-	/* unsigned int			macclone_completed; */
-	struct nat25_network_db_entry	*nethash[NAT25_HASH_SIZE];
-	int				pppoe_connection_in_progress;
-	unsigned char			pppoe_addr[MACADDRLEN];
-	unsigned char			scdb_mac[MACADDRLEN];
-	unsigned char			scdb_ip[4];
-	struct nat25_network_db_entry	*scdb_entry;
-	unsigned char			br_mac[MACADDRLEN];
-	unsigned char			br_ip[4];
+        spinlock_t                          br_ext_lock;
+        /* unsigned int                 macclone_completed; */
+        struct nat25_network_db_entry   *nethash[NAT25_HASH_SIZE];
+        int                             pppoe_connection_in_progress;
+        unsigned char                   pppoe_addr[MACADDRLEN];
+        unsigned char                   scdb_mac[MACADDRLEN];
+        unsigned char                   scdb_ip[4];
+        struct nat25_network_db_entry   *scdb_entry;
+        unsigned char                   br_mac[MACADDRLEN];
+        unsigned char                   br_ip[4];
 
-	struct br_ext_info		ethBrExtInfo;
+        struct br_ext_info              ethBrExtInfo;
     #endif /* CONFIG_BR_SUPPORT */
 #ifdef CONFIG_BAND_STEERING
-	struct workqueue_struct *rsp_wq;
-	struct timer_list steer_timer;
-	struct work_struct steer_work;
-	struct b_steer_priv bsteerpriv;
-	struct ap_probe_rsp pb_pool[MAX_PENDING_PROBES];
+        struct workqueue_struct *rsp_wq;
+        struct timer_list steer_timer;
+        struct work_struct steer_work;
+        struct b_steer_priv bsteerpriv;
+        struct ap_probe_rsp pb_pool[MAX_PENDING_PROBES];
 #endif
 
 };
@@ -554,15 +556,15 @@
 #if (defined CONFIG_HE_FOR_OLD_KERNEL) || (defined CONFIG_VHT_FOR_OLD_KERNEL)
 struct aic_sta {
     u8 sta_idx;            /* Identifier of the station */
-	bool he;               /* Flag indicating if the station supports HE */
+        bool he;               /* Flag indicating if the station supports HE */
     bool vht;               /* Flag indicating if the station supports VHT */
 
-	struct ieee80211_he_cap_elem he_cap_elem;
-	struct ieee80211_he_mcs_nss_supp he_mcs_nss_supp;
+        struct ieee80211_he_cap_elem he_cap_elem;
+        struct ieee80211_he_mcs_nss_supp he_mcs_nss_supp;
 
 #if LINUX_VERSION_CODE < KERNEL_VERSION(3, 8, 0) || defined(CONFIG_VHT_FOR_OLD_KERNEL)
-	__le32 vht_cap_info;
-	struct ieee80211_vht_mcs_info supp_mcs;
+        __le32 vht_cap_info;
+        struct ieee80211_vht_mcs_info supp_mcs;
 #endif
 };
 #endif
@@ -612,15 +614,15 @@
     struct rwnx_sta_stats stats;
     enum nl80211_mesh_power_mode mesh_pm; /*  link-specific mesh power save mode */
 #ifdef CONFIG_DYNAMIC_PERPWR
-	s8_l rssi_save;
-	s8_l per_pwrloss;
-	struct work_struct per_pwr_work;
-	unsigned long last_jiffies;
+        s8_l rssi_save;
+        s8_l per_pwrloss;
+        struct work_struct per_pwr_work;
+        unsigned long last_jiffies;
 #endif
 #ifdef CONFIG_BAND_STEERING
-	u32_l link_time;
-	s8_l rssi;
-	u8_l support_band;
+        u32_l link_time;
+        s8_l rssi;
+        u8_l support_band;
 #endif
 };
 
@@ -727,11 +729,11 @@
 void rwnx_set_conn_state(atomic_t *drv_conn_state, int state);
 
 enum rwnx_drv_connect_status { 
-	RWNX_DRV_STATUS_DISCONNECTED = 0,
-	RWNX_DRV_STATUS_DISCONNECTING, 
-	RWNX_DRV_STATUS_CONNECTING, 
-	RWNX_DRV_STATUS_CONNECTED, 
-	RWNX_DRV_STATUS_ROAMING,
+        RWNX_DRV_STATUS_DISCONNECTED = 0,
+        RWNX_DRV_STATUS_DISCONNECTING, 
+        RWNX_DRV_STATUS_CONNECTING, 
+        RWNX_DRV_STATUS_CONNECTED, 
+        RWNX_DRV_STATUS_ROAMING,
 };
 
 static const char *const s_conn_state[] = {
@@ -744,7 +746,7 @@
 
 
 struct sta_tx_flowctrl {
-	atomic_t tx_pending_cnt;
+        atomic_t tx_pending_cnt;
     u8 flowctrl;
 };
 
@@ -762,7 +764,7 @@
     struct rwnx_vif *vif_table[NX_VIRT_DEV_MAX + NX_REMOTE_STA_MAX]; /* indexed with fw id */
     struct rwnx_sta sta_table[NX_REMOTE_STA_MAX + NX_VIRT_DEV_MAX];
     #ifdef CONFIG_HE_FOR_OLD_KERNEL
-	struct aic_sta aic_table[NX_REMOTE_STA_MAX + NX_VIRT_DEV_MAX];
+        struct aic_sta aic_table[NX_REMOTE_STA_MAX + NX_VIRT_DEV_MAX];
     #endif
     struct rwnx_survey_info survey[SCAN_CHANNEL_MAX];
     struct cfg80211_scan_request *scan_request;
@@ -865,9 +867,9 @@
         bool is_sched_scan;
 #endif//CONFIG_SCHED_SCAN 
 
-	struct sta_tx_flowctrl sta_flowctrl[NX_REMOTE_STA_MAX];
+        struct sta_tx_flowctrl sta_flowctrl[NX_REMOTE_STA_MAX];
 #if 0
-	bool he_flag;
+        bool he_flag;
 #endif
 #if LINUX_VERSION_CODE < KERNEL_VERSION(3, 6, 0)
     struct mac_chan_op ap_chan;
@@ -879,23 +881,23 @@
 #endif
 
 #ifdef CONFIG_USE_WIRELESS_EXT
-	bool wext_scan;
-	struct completion wext_scan_com;
-	struct list_head wext_scanre_list;
-	char wext_essid[33];
-	int support_freqs[SCAN_CHANNEL_MAX];
-	int support_freqs_number;
+        bool wext_scan;
+        struct completion wext_scan_com;
+        struct list_head wext_scanre_list;
+        char wext_essid[33];
+        int support_freqs[SCAN_CHANNEL_MAX];
+        int support_freqs_number;
 #ifdef CONFIG_DYNAMIC_PWR
-	struct timer_list pwrloss_timer;
-	struct work_struct pwrloss_work;
-	struct rwnx_vif *read_rssi_vif;
-	s8 pwrloss_lvl;
-	u8 sta_rssi_idx;
+        struct timer_list pwrloss_timer;
+        struct work_struct pwrloss_work;
+        struct rwnx_vif *read_rssi_vif;
+        s8 pwrloss_lvl;
+        u8 sta_rssi_idx;
 #endif
 #endif
 #ifdef CONFIG_BAND_STEERING
-	u8_l iface_idx;
-	struct tmp_feature_sta feature_table[NX_REMOTE_STA_MAX + NX_VIRT_DEV_MAX];
+        u8_l iface_idx;
+        struct tmp_feature_sta feature_table[NX_REMOTE_STA_MAX + NX_VIRT_DEV_MAX];
 #endif
 };
 
@@ -910,13 +912,13 @@
 static inline bool is_multicast_sta(int sta_idx)
 {
 
-	if((g_rwnx_plat->usbdev->chipid == PRODUCT_ID_AIC8801) || 
-		((g_rwnx_plat->usbdev->chipid == PRODUCT_ID_AIC8800DC ||
-		g_rwnx_plat->usbdev->chipid == PRODUCT_ID_AIC8800DW) && chip_id < 3)){
-		return (sta_idx >= NX_REMOTE_STA_MAX_FOR_OLD_IC);
-	}else{
-		return (sta_idx >= NX_REMOTE_STA_MAX);
-	}
+        if((g_rwnx_plat->usbdev->chipid == PRODUCT_ID_AIC8801) || 
+                ((g_rwnx_plat->usbdev->chipid == PRODUCT_ID_AIC8800DC ||
+                g_rwnx_plat->usbdev->chipid == PRODUCT_ID_AIC8800DW) && chip_id < 3)){
+                return (sta_idx >= NX_REMOTE_STA_MAX_FOR_OLD_IC);
+        }else{
+                return (sta_idx >= NX_REMOTE_STA_MAX);
+        }
 
 }
 struct rwnx_sta *rwnx_get_sta(struct rwnx_hw *rwnx_hw, const u8 *mac_addr);
