--- a/drivers/aic8800/aic_load_fw/aic_compat_8800d80.c	2026-05-19 05:53:22.611233508 +0000
+++ b/drivers/aic8800/aic_load_fw/aic_compat_8800d80.c	2026-05-19 05:53:54.137060947 +0000
@@ -337,6 +337,7 @@
 }
 
 
+#if 0
 static int aicbt_ext_patch_data_load(struct aic_usb_dev *usb_dev, struct aicbt_patch_info_t *patch_info)
 {
     int ret = 0;
@@ -369,8 +370,10 @@
 }
 
 
+#endif
 int aicfw_download_fw_8800d80(struct aic_usb_dev *usb_dev)
 {
+#if 0
     struct aicbt_patch_table *head = NULL;
     struct aicbt_patch_info_t patch_info = {
         .info_len          = 0,
@@ -383,6 +386,7 @@
         .adid_flag_addr    = 0,
         .adid_flag         = 0,
     };
+#endif
 
     int i = 0;
 
