--- a/drivers/aic8800/aic_load_fw/aicbluetooth.c
+++ b/drivers/aic8800/aic_load_fw/aicbluetooth.c
@@ -757,7 +757,7 @@ int rwnx_plat_flash_bin_upload_android(struct aic_usb_dev *usbdev, u32 fw_addr,
     u32 *dst=NULL;
     int err=0;
     const u32 mem_addr = fw_addr;
     struct dbg_mem_read_cfm rd_mem_addr_cfm;
-    u32 crc = ~0UL;
+    u32 crc = ~0U;

     /* load aic firmware */
     size = aic_load_firmware(&dst, filename, dev);
--- a/drivers/aic8800/aic8800_fdrv/rwnx_msg_rx.c
+++ b/drivers/aic8800/aic8800_fdrv/rwnx_msg_rx.c
@@ -815,7 +815,9 @@ static inline int rwnx_rx_scanu_result_ind(struct rwnx_hw *rwnx_hw,
 #endif

     }
+#ifdef CONFIG_USE_WIRELESS_EXT
 putbss:
+#endif
     if (bss != NULL)
 #if LINUX_VERSION_CODE < KERNEL_VERSION(3, 9, 0)
         cfg80211_put_bss(bss);

